### PR TITLE
Feat/remote file

### PR DESCRIPTION
**Pull request by jaroug**
Add support for probing a remote PEM file over http/https

This is a corner case to be able to monitor BIMI related certificates

- https://support.google.com/a/answer/10911320
- https://support.google.com/a/answer/10911321

### CODE REVIEW

**Comment by dragoangel** [comment]
Looks nice

**Review by ribbybibby — CHANGES_REQUESTED** [review]
Thank you for this PR and sorry for taking so long to review it!

I have a few comments

1. I don't think this should export `ssl_file_cert_*` metrics. We should just read the certificates out of the response body and export the regular `ssl_cert_*` metrics. We don't need the intermediate step where we write them to disk.
2. This also exports metrics from the TLS connection because it reuses the tls config used in the https prober: https://github.com/ribbybibby/ssl_exporter/blob/42d1d4ff0f0b08e405457277eb83460d16c278d1/prober/tls.go#L30. I don't think it should do this. It should only export metrics for the certificates returned in the response body.
3. Can you please improve the tests so that they actually verify the exported metrics?

**Comment by jaroug** [comment]
Hi :) no worries I'm glad the project is still alive.

I made the changes requested, at least I hope it matches your expectations. I honestly can't remember why I thought it was a good idea to download the file :sweat_smile:

**Comment by dragoangel** [comment]
LGTM

**Comment by dragoangel** [comment]
@ribbybibby can you please review if this PR now is okay?

File: prober/remote_file.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/ribbybibby/ssl_exporter/v2/config"
+)
+
+// ProbeFile collects certificate metrics from local files
```
Comment (ribbybibby) [inline prober/remote_file.go:14]:
```suggestion
// ProbeFile collects certificate metrics from a remote file
```

File: prober/remote_file.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			errCh <- fmt.Errorf("failed to download file, status code: %d", resp.StatusCode)
+			return
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		certs, err := decodeCertificates([]byte(body))
```
Comment (ribbybibby) [inline prober/remote_file.go:51]: These errors should be handled.

File: README.md [hunk 3]
```diff
@@ -175,6 +176,39 @@ scrape_configs:
         replacement: ${1}:9219
 ```
 
+### Remote_file
+
+The `remote_file` prober exports `ssl_file_cert_not_after` and
+`ssl_file_cert_not_before` for PEM encoded certificates found at the
+specified remote target.
```
Comment (ribbybibby) [inline README.md:183]:
```suggestion
The `remote_file` prober exports `ssl_cert_not_after` and
`ssl_cert_not_before` for PEM encoded certificates found at the
specified remote target.
```

**Review by ribbybibby — CHANGES_REQUESTED** [review]
Changes suggested/requested in comments.

**Comment by ribbybibby** [comment]
Thank you for the contribution @jaroug. I've made some of my suggested changes and fixed some other things up and now I think this is good to merge.

**Comment by dragoangel** [comment]
@ribbybibby can we get 2.5.0 released with this feature please? 🙏

**Comment by dragoangel** [comment]
@ribbybibby kind reminder - this feature not get into 2.4.3 release, as it was been released before merging this branch.

**Comment by dragoangel** [comment]
@jaroug btw, I found out that `ssl_verified_cert_not_after` not available as metrics for `http_file`, is this desired behaviour?

**Comment by cHiv0rz** [comment]
@ribbybibby do you have plans to create a new release including this functionality?

**Comment by dragoangel** [comment]
@cHiv0rz I also was waiting for it, but ended up building own image and pushing it to my registry.

**Comment by cHiv0rz** [comment]
Yes, mee to, compiling from master and uploading the image. But just wanted to know if the exporter will be maintained or not :D

**Comment by cHiv0rz** [comment]
@dragoangel do you use this functionality for BIMI certificates? 

I'm getting the following error:

```sh
ts=2025-05-12T10:08:16.013Z caller=ssl_exporter.go:107 level=error target=https://some/target/file.pem prober=http_file timeout=15s msg="making http request: Get \"https://some/target/file.pem": tls: failed to verify certificate: x509: certificate signed by unknown authority"
```

The certificate was issued by DigiCert, that I think is known abroad 😓 Have you faced this problem?

**Comment by cHiv0rz** [comment]
nevermind, it was my fault not adding the certificates to the scratch container

**Comment by dragoangel** [comment]
Yep, using and it works good

**Pull request merged**